File: 大一上/图灵班/程序设计荣誉课程-1/yoj全解/YOJ-90/a.py
'''
Author: zmlad
Date: 2025-12-09 14:30:31
'''
def main():
    n = int(input())
    required = list(map(int, input().split()))
    
    m, k = map(int, input().split())
    
    students = []
    
    for _ in range(m):
        data = input().split()
        sno = int(data[0])
        p = int(data[1])
        solved_problems = list(map(int, data[2:2+p]))
        
        solved = 0
        counted = [False] * n  # 记录是否已经统计过该必做题
        
        for pid in solved_problems:
            for r in range(n):
                if pid == required[r] and not counted[r]:
                    counted[r] = True
                    solved += 1
                    break
        
        students.append((sno, solved))
    
    # 排序：解决题目数降序，学号升序
    students.sort(key=lambda x: (-x[1], x[0]))
    
    # 密集排名并输出前k名
    rank = 1
    prev_solved = students[0][1]
    result = []
    
    for i in range(len(students)):
        if i > 0 and students[i][1] != prev_solved:
            rank += 1
            prev_solved = students[i][1]
        
        if rank > k:
            break
        
        result.append(str(students[i][0]))
    
    print(" ".join(result))

if __name__ == "__main__":
    main()

File: 大一上/图灵班/程序设计荣誉课程-1/yoj全解/YOJ-90/a.cpp
/*
 * @Author: zmlad
 * @Date: 2025-12-09 14:30:20
 */
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

struct Student {
    long long sno;
    int solved;
};

// 比较函数：按解决题目数降序，学号升序
bool cmp(const Student &a, const Student &b) {
    if (a.solved != b.solved) return a.solved > b.solved;
    return a.sno < b.sno;
}

int main() {
    int n;
    cin >> n;
    
    vector<int> required(n);
    for (int i = 0; i < n; i++) {
        cin >> required[i];
    }
    
    int m, k;
    cin >> m >> k;
    
    vector<Student> students(m);
    
    for (int idx = 0; idx < m; idx++) {
        long long sno;
        int p;
        cin >> sno >> p;
        
        int solved = 0;
        vector<bool> counted(n, false); // 记录是否已经统计过该必做题
        
        for (int j = 0; j < p; j++) {
            int pid;
            cin >> pid;
            // 检查是否是必做题且未统计过
            for (int r = 0; r < n; r++) {
                if (pid == required[r] && !counted[r]) {
                    counted[r] = true;
                    solved++;
                    break;
                }
            }
        }
        
        students[idx].sno = sno;
        students[idx].solved = solved;
    }
    
    // 排序
    sort(students.begin(), students.end(), cmp);
    
    // 密集排名并输出前k名
    int rank = 1;
    int prev_solved = students[0].solved;
    
    for (int i = 0; i < m; i++) {
        if (i > 0 && students[i].solved != prev_solved) {
            rank++;
            prev_solved = students[i].solved;
        }
        
        if (rank > k) break;
        
        if (i > 0) cout << " ";
        cout << students[i].sno;
    }
    
    cout << endl;
    return 0;
}